### PR TITLE
Add element data client trust options.

### PR DESCRIPTION
**Pull request by TheNormalnij**
Added `elementdata_whitelisted` setting for `mtaserver.conf`
* Enabled: the client can't change element data on server, except allowed keys.
* Disabled: the client can change any element data. (default, unsafe)

Added `clientTrustMode` parameter for` setElementData`

```lua
-- Allow changes from the client
bool setElementData(root,  key, value, syncType, "allow" )

-- Deny changes from the client
bool setElementData(root,  key, value, syncType, "deny" )

-- Respect elementdata_whitelisted setting
bool setElementData(root,  key, value, syncType, "default" )
```

Added event `onPlayerChangesProtectedData`
source: player
arguments:
* element,
* key,
* value

### CODE REVIEW

**Comment by TracerDS** [comment]
Whats the purpose of these functions?

File: Server/mods/deathmatch/logic/CCustomData.h [hunk 1]
```diff
@@ -25,23 +25,37 @@ enum class ESyncType
     SUBSCRIBE,
 };
 
+enum class ECustomDataClientTrust : std::uint8_t
```
Comment (TracerDS) [inline Server/mods/deathmatch/logic/CCustomData.h]: lowercase e? (`eCustomDataClientTrust`)

File: Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp [hunk 2]
```diff
@@ -2437,3 +2443,27 @@ int CLuaElementDefs::isElementCallPropagationEnabled(lua_State* luaVM)
...
+
+bool CLuaElementDefs::IsElementDataClientTrusted(CElement* pElement, std::optional<std::string_view> key)
+{
+    if (key.has_value())
+        return pElement->GetCustomDataManager().IsClientChangesAllowed(key.value().data());
+    else
+        return pElement->GetCustomDataManager().IsClientChangesAllowed();
+}
+
+void CLuaElementDefs::ResetElementDataClientTrust(CElement* pElement, std::optional<std::string_view> key)
```
Comment (TracerDS) [inline Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: maybe noexcept?

**Review by TracerDS — CHANGES_REQUESTED** [review]

**Comment by ffsPLASMA** [comment]
isnt this kinda redundant with the way you can already detect elementdata changes in lua? thats why I added it to security resource.

https://github.com/multitheftauto/mtasa-resources/blob/master/%5Badmin%5D/security/players.lua#L9

**Comment by FileEX** [comment]
From what I understand, the issue is that if we have an elementData key, like "user:points", which we only use in server-side scripts and it’s not synchronized on the server, a potential cheater could still overwrite it by using setElementData client-side with the sync parameter set to true. My suggestion is to add a new server-side argument, something like ignoreClientSync. Then, in the server’s elementData synchronization packet, if a particular elementData key has the ignoreClientSync parameter set to true, this synchronization packet would simply be ignored, and the value would remain unchanged.

In my opinion, a single simple argument is a much better solution than having entirely separate functions for this

**Comment by ffsPLASMA** [comment]
> From what I understand, the issue is that if we have an elementData key, like "user:points", which we only use in server-side scripts and it’s not synchronized on the server, a potential cheater could still overwrite it by using setElementData client-side with the sync parameter set to true. My suggestion is to add a new server-side argument, something like ignoreClientSync. Then, in the server’s elementData synchronization packet, if a particular elementData key has the ignoreClientSync parameter set to true, this synchronization packet would simply be ignored, and the value would remain unchanged.
> 
> In my opinion, a single simple argument is a much better solution than having entirely separate functions for this

Ye I agree something like:

bool setElementData ( element theElement, string key, var value [, bool synchronize = true ], [bool acceptClientUpdate = true] )

If synchronize is set to false, acceptClientUpdate should also be set to false.

**Comment by PlatinMTA** [comment]
> From what I understand, the issue is that if we have an elementData key, like "user:points", which we only use in server-side scripts and it’s not synchronized on the server, a potential cheater could still overwrite it by using setElementData client-side with the sync parameter set to true. My suggestion is to add a new server-side argument, something like ignoreClientSync. Then, in the server’s elementData synchronization packet, if a particular elementData key has the ignoreClientSync parameter set to true, this synchronization packet would simply be ignored, and the value would remain unchanged.
> 
> In my opinion, a single simple argument is a much better solution than having entirely separate functions for this

What if the elementdata is not set serverside? Should it be ignored by default? I do believe this should be the case but I also think its too late to be making these changes...

**Comment by FileEX** [comment]
> 
> What if the elementdata is not set serverside? Should it be ignored by default? I do believe this should be the case but I also think its too late to be making these changes...

If an element data doesn’t exist on the server side for a given element, like player:vip, the scripter should simply set this element data to false by default rather than leaving it unset. If any element data isn’t used at all on the server side, it doesn’t matter if a potential cheater sets it. Yes, there’s still the issue that a cheater could send a packet manually, bypassing all checks, but I’ve already suggested a potential solution for this to Dutchman. In any case, an argument like acceptClientUpdate would make manipulating element data with executors impossible, as the server would ignore that packet

**Comment by wassmas** [comment]
> From what I understand, the issue is that if we have an elementData key, like "user:points", which we only use in server-side scripts and it’s not synchronized on the server, a potential cheater could still overwrite it by using setElementData client-side with the sync parameter set to true. My suggestion is to add a new server-side argument, something like ignoreClientSync. Then, in the server’s elementData synchronization packet, if a particular elementData key has the ignoreClientSync parameter set to true, this synchronization packet would simply be ignored, and the value would remain unchanged.
> 
> In my opinion, a single simple argument is a much better solution than having entirely separate functions for this

Thats why i request for this: #3696

**Comment by TheNormalnij** [comment]
> it doesn’t matter if a potential cheater sets it

Nope. The data will be synced with all clients. 

> What if the elementdata is not set serverside? Should it be ignored by default?

I'll add a server setting to ignore all unknown element data from clients.

File: Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp [hunk 3]
```diff
@@ -984,7 +989,7 @@ bool CStaticFunctionDefinitions::SetElementData(CElement* pElement, const char*
         // Unsubscribe all the players
         if (lastSyncType == ESyncType::SUBSCRIBE && syncType != ESyncType::SUBSCRIBE)
             m_pPlayerManager->ClearElementData(pElement, szName);
-
+
```
Comment (TheNormalnij) [inline Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp]:
```suggestion

```

File: Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp [hunk 4]
```diff
@@ -984,7 +989,6 @@ bool CStaticFunctionDefinitions::SetElementData(CElement* pElement, const char*
         // Unsubscribe all the players
         if (lastSyncType == ESyncType::SUBSCRIBE && syncType != ESyncType::SUBSCRIBE)
             m_pPlayerManager->ClearElementData(pElement, szName);
```
Comment (TheNormalnij) [inline Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp]:
```suggestion
            m_pPlayerManager->ClearElementData(pElement, szName);

```

**Comment by ffsPLASMA** [comment]
Wouldnt it be better to just outright dismiss all client sync updates to an elementdata if the data wasnt shared to client in first place?

eg. setElementData(root, "coolData", "mreow", false)

^doesnt get synced to client, so why not disregard any client updates for said key in first place?

**Comment by TheNormalnij** [comment]
> Wouldnt it be better to just outright dismiss all client sync updates to an elementdata if the data wasnt shared to client in first place?
> 
> eg. setElementData(root, "coolData", "mreow", false)
> 
> ^doesnt get synced to client, so why not disregard any client updates for said key in first place?

It is not enough. Element data may be broadcasted by server only. In this case the element data is shared, but the server deny all changes from client. Whitelisting is the safest option.

**Comment by ffsPLASMA** [comment]
> > Wouldnt it be better to just outright dismiss all client sync updates to an elementdata if the data wasnt shared to client in first place?
> > eg. setElementData(root, "coolData", "mreow", false)
> > ^doesnt get synced to client, so why not disregard any client updates for said key in first place?
> 
> It is not enough. Element data may be broadcasted by server only. In this case the element data is shared, but the server deny all changes from client. Whitelisting is the safest option.

I dont quite understand.

When I create an elementdata on server and specifically set it to not get synced to any clients, then the server should not accept/retrieve any updates from client for said key in first place. If I explicitly set an elementdata to get synced to other clients, I can use onElementDataChange to see if said client is allowed to change it or not.

Adding yet another parameter to setElementData to pay attention to and a server config setting is kinda nonsense.

**Comment by TheNormalnij** [comment]
I agree that local serverside element data may be protected by default.
But broadcasted element data requests protection too

**Comment by Dutchman101** [comment]
Thanks!
Regarding desirability: the PR has been requested by AC team. You don't have to use it if you don't want to.

**Comment by Fernando-A-Rocha** [comment]
Cool, script security wiki article should be updated

**Pull request merged**